imqsauth.go

```diff
@@ -234,6 +251,221 @@ func createDB(config *authaus.Config) bool {
+	} else {
+		fmt.Fprintf(w, "\tPermit list ok:\t%v permits\n", len(plist))
+		fmt.Fprintf(w, "\t*Nr of non-archived users without permits:\t%v\n", stats[unArchivedWithoutPermit])
+		fmt.Fprintf(w, "\t*Nr of archived users without permits:\t%v\n", len(users)-len(plist)-stats[unArchivedWithoutPermit])
+
+	}
+
+	// Non-existent groups
+	if len(missingGroups) > 0 {
+		fmt.Fprintf(w, "Missing groups:\n")
+	}
+
+	groupKeys := maps.Keys(missingGroups)
+	slices.SortFunc(groupKeys, func(a authaus.GroupIDU32, b authaus.GroupIDU32) int { return int(a) - int(b) })
+	for _, v := range groupKeys {
+		slices.SortFunc(missingGroups[v], func(a authaus.UserId, b authaus.UserId) int { return int(a) - int(b) })
+		fmt.Fprintf(w, "\tGroup missing:\t%v\tUsers: %v\n", v, missingGroups[v])
+	}
+
+	fmt.Fprintf(w, "\n* - indicates possibly inaccurate values.\n")
+
+	// Errors
+	fmt.Fprintln(w, "")
+	fmt.Fprintln(w, "Errors (explicit only - review the missing group list for all missing groups)")
+	fmt.Fprintln(w, "------")
+
+	errKeys := maps.Keys(errorSet)
+	sort.Strings(errKeys)
+
+	if len(errKeys) > 0 {
+		for _, e := range errKeys {
+			fmt.Fprintf(w, "%v\n", errorSet[e])
+		}
+	} else {
+		fmt.Fprintf(w, "No errors\n")
+	}
+
+	// FINAL HEALTHCHECK OUTPUT
+	w.Flush()
+	fmt.Print(sb.String())
+
+	// CLEAN MISSING GROUPS IF REQUESTED
+	if cleanAll {
+		for _, g := range groupKeys {
+			cleanGroup(ic, strconv.Itoa(int(g)))
+		}
+	}
+
+	return true
+}
+
+func getUserStats(ic *auth.ImqsCentral, users []authaus.AuthUser, stats map[string]int, typeCounts map[authaus.AuthUserType]*enabledTotal, errorSet map[string]error) (map[authaus.GroupIDU32][]authaus.UserId, bool, error) {
+	var err error
+	errPerms := false
+	missingGroups := map[authaus.GroupIDU32][]authaus.UserId{}
+	for _, u := range users {
+		p, _ := ic.Central.GetPermit(u.UserId)
+
+		if u.Archived {
+			stats[archived]++
+		}
+
+		if p == nil && !u.Archived {
+			stats[unArchivedWithoutPermit]++
+		}
+
+		typeCounts[u.Type].total++
+		if perm, ePerm := ic.Central.GetPermit(u.UserId); ePerm == nil {
+			// cycle through permit and check if group exists
+			if ePerm == nil {
+				groupIds, _ := authaus.DecodePermit(perm.Roles)
+				for _, r := range groupIds {
+					if _, err := ic.Central.GetRoleGroupDB().GetByID(r); err != nil {
+						missingGroups[r] = append(missingGroups[r], u.UserId)
+					}
+
+				}
+			}
+			pbits, eGroup := authaus.PermitResolveToList(perm.Roles, ic.Central.GetRoleGroupDB())
+			if pbits != nil {
+				if eGroup != nil {
+					errPerms = true
+					errorSet[eGroup.Error()] = eGroup
+				}
+				if pbits.Has(auth.PermEnabled) {
+					stats[enabled]++
+					typeCounts[u.Type].enabled++
+				}
+				if pbits.Has(auth.PermAdmin) {
+					stats[admin]++
+				}
+			} else {
+				errPerms = true
+			}
+		} else {
+			err = ePerm
+		}
+	}
+	return missingGroups, errPerms, err
+}
+
+func findMissingOAuthTokens(oAuthTokenIDs []string, tokens []*authaus.Token) int {
+	oAuthIDs := map[string]struct{}{}
```

Comment:
```suggestion
	oAuthIDs := make(map[string]bool)
```

Reply:
https://itnext.io/set-in-go-map-bool-and-map-struct-performance-comparison-5315b4b107b

Reply:
Dang... I even added some more realistic use cases, and it still won... I guess maps have more overhead than I thought :P and I now have to go back and change a bunch of code...